### PR TITLE
remember

### PR DESCRIPTION
**Pull request by MinhDo2007**

### CODE REVIEW

File: app/controllers/sessions_controller.rb [hunk 1]
```diff
@@ -3,18 +3,21 @@ def new
   end
 
   def create
-    user = User.find_by email: params[:session][:email].downcase
+    user = User.find_by(email: params[:session][:email].downcase)
```
Comment (phamkykhoi) [inline app/controllers/sessions_controller.rb]: Remove ( ) in line User.find_by(email: params[:session][:email].downcase)

File: app/controllers/sessions_controller.rb [hunk 2]
```diff
@@ -3,18 +3,21 @@ def new
...
     if user && user.authenticate(params[:session][:password])
       log_in user
-      redirect_back_or user
+      remember user
+      params[:session][:remember_me] == '1' ? remember(user) : forget(user)
+      redirect_to user
     else
-      flash[:error] = t "login.failed"
-      render "new" 
+      flash[:error] = t("login.failed")
```
Comment (phamkykhoi) [inline app/controllers/sessions_controller.rb]: Remove ( )

File: app/helpers/sessions_helper.rb [hunk 3]
```diff
@@ -1,15 +1,30 @@
...
 
+  def remember user
+    user.remember
+    cookies.permanent.signed[:user_id] = user.id
+    cookies.permanent[:remember_token] = user.remember_token
+  end
+
   def current_user
-    @current_user ||= User.find_by id: session[:user_id]
+    if (user_id = session[:user_id])
```
Comment (phamkykhoi) [inline app/helpers/sessions_helper.rb]: Remove ( )

File: app/helpers/sessions_helper.rb [hunk 4]
```diff
@@ -1,15 +1,30 @@
...
+      @current_user ||= User.find_by(id: user_id)
```
Comment (phamkykhoi) [inline app/helpers/sessions_helper.rb]: Remove ( )

File: app/helpers/sessions_helper.rb [hunk 5]
```diff
@@ -1,15 +1,30 @@
...
+    elsif (user_id = cookies.signed[:user_id])
+      user = User.find_by(id: user_id)
```
Comment (phamkykhoi) [inline app/helpers/sessions_helper.rb]: Remove ( )

File: app/helpers/sessions_helper.rb [hunk 6]
```diff
@@ -36,4 +51,16 @@ def logged_in_user
       redirect_to login_url
     end
   end
+
+  def forget user
+    user.forget
+    cookies.delete(:user_id)
```
Comment (phamkykhoi) [inline app/helpers/sessions_helper.rb]: Remove ()

File: app/helpers/sessions_helper.rb [hunk 7]
```diff
@@ -36,4 +51,16 @@ def logged_in_user
...
+    cookies.delete(:remember_token)
```
Comment (phamkykhoi) [inline app/helpers/sessions_helper.rb]: Remove ()

File: app/helpers/sessions_helper.rb [hunk 8]
```diff
@@ -36,4 +51,16 @@ def logged_in_user
...
+  end
+
+  def log_out
+    forget(current_user)
```
Comment (phamkykhoi) [inline app/helpers/sessions_helper.rb]: Remove ()

File: app/helpers/sessions_helper.rb [hunk 9]
```diff
@@ -36,4 +51,16 @@ def logged_in_user
...
+    session.delete(:user_id)
```
Comment (phamkykhoi) [inline app/helpers/sessions_helper.rb]: Remove ( )

File: app/models/user.rb [hunk 10]
```diff
@@ -27,4 +31,22 @@ def User.digest value
...
     BCrypt::Password.create(value, cost: cost)
   end
+
+  def User.new_token
+    SecureRandom.urlsafe_base64
+  end
+
+  def remember
+    self.remember_token = User.new_token
+    update_attributes remember_digest: User.digest(remember_token)
```
Comment (phamkykhoi) [inline app/models/user.rb]: Remove ( )

File: app/models/user.rb [hunk 11]
```diff
@@ -27,4 +31,22 @@ def User.digest value
     cost = ActiveModel::SecurePassword.min_cost if BCrypt::Engine::MIN_COST
     BCrypt::Password.create(value, cost: cost)
```
Comment (phamkykhoi) [inline app/models/user.rb]: Remove ( )

File: app/models/user.rb (hunk 10 above)
Comment (phamkykhoi) [inline app/models/user.rb]: Remove ( )

Comment (MinhDo2007) [inline app/models/user.rb]: I don't remove () in update_attributes remember_digest: User.digest(remember_token)

File: app/helpers/sessions_helper.rb [hunk 12]
```diff
@@ -1,15 +1,30 @@
...
+    cookies.permanent.signed[:user_id] = user.id
+    cookies.permanent[:remember_token] = user.remember_token
+  end
+
   def current_user
-    @current_user ||= User.find_by id: session[:user_id]
+    if user_id = session[:user_id]
+      @current_user ||= User.find_by id: user_id
+    elsif user_id = cookies.signed[:user_id]
+      user = User.find_by id: user_id
```
Comment (phamkykhoi) [inline app/helpers/sessions_helper.rb]: User.find_by id: user_id => User.find user_id

File: app/helpers/sessions_helper.rb [hunk 13]
```diff
@@ -1,15 +1,30 @@
...
+  def remember user
+    user.remember
+    cookies.permanent.signed[:user_id] = user.id
+    cookies.permanent[:remember_token] = user.remember_token
+  end
+
   def current_user
-    @current_user ||= User.find_by id: session[:user_id]
+    if user_id = session[:user_id]
+      @current_user ||= User.find_by id: user_id
```
Comment (phamkykhoi) [inline app/helpers/sessions_helper.rb]: User.find_by id: user_id => User.find user_id

File: app/models/user.rb [hunk 14]
```diff
@@ -1,7 +1,7 @@
 class User < ActiveRecord::Base
   before_create :create_activation_digest
-
-  attr_accessor :activation_token
+
```
Comment (TrucLuu) [inline app/models/user.rb]: remove empty line

File: app/controllers/sessions_controller.rb [hunk 15]
```diff
@@ -6,15 +6,18 @@ def create
     user = User.find_by email: params[:session][:email].downcase
     if user && user.authenticate(params[:session][:password])
       log_in user
-      redirect_back_or user
+      remember user
+      params[:session][:remember_me] == '1' ? remember(user) : forget(user)
```
Comment (dieunb) [inline app/controllers/sessions_controller.rb]: `"` not `'`

File: app/helpers/sessions_helper.rb [hunk 16]
```diff
@@ -1,15 +1,30 @@
 module SessionsHelper
-  def log_in user
+  def log_in user
```
Comment (dieunb) [inline app/helpers/sessions_helper.rb]: remove trailing spaces

File: app/models/user.rb (hunk 14 above)
Comment (dieunb) [inline app/models/user.rb]: remove trailing spaces

**Comment by dieunb** [comment]
OK

**Pull request merged**